src/lib/raft/raft.c

```diff
@@ -394,6 +455,10 @@ raft_process_msg(struct raft *raft, const struct raft_msg *req, uint32_t source)
 	/* Term bump. */
 	if (req->term > raft->volatile_term)
 		raft_sm_schedule_new_term(raft, req->term);
+
+	/* Notification from a remote node that it sees the current leader. */
+	raft_notify_leader_seen(raft, req->leader != 0, source);
```

Comment:
> Otherwise you have 4 places which make a broadcast when this function returns true, and this is the only place which doesn't.

That's because that 4 places are with `raft->self` as a source. We should broadcast the fact that this node sees (or doesn't see) the leader.
We don't care when someone else (or noone) sees the leader though. Hence no broadcast in this particular case.

Sounds like `raft_notify_leader_seen` with `raft->self` as a source should be a separate function, then, and should have a broadcast every time `leader_seen` is cahnged. But I couldn't come up with an appropriate name or semantics. I already have `raft_set_leader` and `raft_clear_leader`. It could be `raft_leader_appear` and `raft_leader_disappear`, maybe?

> Besides, shouldn't you call raft_sm_wait_leader_dead() if the witness map was 0 but became != 0 here?
No. The timer is only active as long as this node sees the leader directly. And the timeout means that the node doesn't see the leader anymore. So I can't reset the timer when someone else sees the leader but this node doesn't.

Reply:
> That's because that 4 places are with raft->self as a source. We should broadcast the fact that this node sees (or doesn't see) the leader. We don't care when someone else (or noone) sees the leader though. Hence no broadcast in this particular case.

Yes, I see now, thanks for the explanation.

> Sounds like raft_notify_leader_seen with raft->self as a source should be a separate function, then, and should have a broadcast every time leader_seen is cahnged. But I couldn't come up with an appropriate name or semantics.

Could also add some branching into this function and handle both self and not self. See my proposals in the new review. The idea with 2 separate functions also sounds fine for me, at least theoretically. The names are fine. Other possible option: `raft_leader_see`/`raft_leader_unsee`.

Reply:
I've implemented the idea with `raft_leader_see / raft_leader_unsee`. Please see the diff in reply to your new review.